api/src/main/java/io/druid/data/input/impl/prefetch/PrefetchSqlFirehoseFactory.java

```diff
@@ -0,0 +1,253 @@
+import io.druid.java.util.common.ISE;
+import io.druid.java.util.common.concurrent.Execs;
+import io.druid.java.util.common.logger.Logger;
+import com.fasterxml.jackson.databind.ObjectMapper;
+import org.apache.commons.io.LineIterator;
+
+import javax.annotation.Nullable;
+import java.io.File;
+import java.io.IOException;
+import java.io.InputStream;
+import java.util.Collection;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+import java.util.NoSuchElementException;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.TimeUnit;
+
+/**
+ * PrefetchSqlFirehoseFactory is an abstract firehose factory for reading prefetched sql resultset data. Regardless
+ * of whether prefetching is enabled or not, for each sql object the entire result set is fetched into a file in the local disk.
+ * This class defines prefetching as caching the resultsets into local disk in case multiple sql queries are present.
+ * When prefetching is enabled, the following functionalities are provided:
+ * <p/>
+ * <p>
+ * - Caching: for the first call of {@link #connect(InputRowParser, File)}, it caches objects in a local disk
+ * up to maxCacheCapacityBytes.  These caches are NOT deleted until the process terminates, and thus can be used for
+ * future reads.
+ * <br/>
+ * - Fetching: when it reads all cached data, it fetches remaining objects into a local disk and reads data from
+ * them.  For the performance reason, prefetch technique is used, that is, when the size of remaining fetched data is
+ * smaller than {@link PrefetchConfig#prefetchTriggerBytes}, a background prefetch thread automatically starts to fetch remaining
+ * objects.
+ * <br/>
+ * <p/>
+ * <p>
+ * This implementation aims to avoid maintaining a persistent connection to the database by prefetching the resultset into disk.
+ * <br/>
+ * Prefetching can be turned on/off by setting maxFetchCapacityBytes.  Depending on prefetching is enabled or
+ * disabled, the behavior of the firehose is different like below.
+ * <p/>
+ * <p>
+ * 1. If prefetch is enabled this firehose can fetch input objects in background.
+ * <br/>
+ * 2. When next() is called, it first checks that there are already fetched files in local storage.
+ * <br/>
+ * 2.1 If exists, it simply chooses a fetched file and returns a {@link LineIterator} reading that file.
+ * <br/>
+ * 2.2 If there is no fetched files in local storage but some objects are still remained to be read, the firehose
+ * fetches one of input objects in background immediately. Finally, the firehose returns an iterator of {@link JsonIterator}
+ * for deserializing the saved resultset.
+ * <br/>
+ * 3. If prefetch is disabled, the firehose saves the resultset to file and returns an iterator of {@link JsonIterator}
+ * which directly reads the stream opened by {@link #openObjectStream}. If there is an IOException, it will throw it
+ * and the read will fail.
+ */
+public abstract class PrefetchSqlFirehoseFactory<T>
```

Comment:
It looks like most code is duplicated with `PrefetchableFirehoseFactory` which might cause the code maintenance issue. Is it possible to reduce it? I guess it's possible if the prefetching is abstracted to a super class and its subclasses specify the object types (like database tables or text files) they handle.

Reply:
@jihoonson Thanks for reviewing. The prefetching logic has already been abstracted into the `Fetcher` class. `PrefetchSqlFirehoseFactory`  is laid out similar to `PrefetchableTextFilesFirehoseFactory` but use different implementations of `Fetcher` and `Firehose`.
Since `PrefetchableTextFilesFirehoseFactory` is already an extension of  `AbstractTextFilesFirehoseFactory` and is dependent on `StringInputRowParser` and since it is not the same case for `PrefetchSqlFirehoseFactory` , I thought it made sense to abstract out just the prefetching logic. Please let me know what you think.